### PR TITLE
feat(auth): use the viewer endpoint to load roles array

### PR DESCRIPTION
**Pull request by joelhooks**
rolls

### CODE REVIEW

File: src/lib/viewer.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import {getGraphQLClient} from '../utils/configured-graphql-client'
+import gql from 'graphql-tag'
+
+export async function loadCurrentViewerRoles(token: string) {
+  const query = gql`
+    query viewer {
+      viewer {
```
Comment (joelhooks) [inline src/lib/viewer.ts:7]: this is an alias server side in gql for current user and generally in egghead-next we've tried to use "viewer" terminology over "user" with varied levels of success

the idea being that a "viewer" isn't necessarily the user but could be support viewing as the user, as a for instance

that's why we have a ViewerProvider etc

Comment (joelhooks) [reply src/lib/viewer.ts:7]: Figured gql just grabbing the roles array would be quicker than jbuilder any day of the week

File: src/server/ability.ts [hunk 2]
```diff
@@ -1,42 +1,42 @@
 import {AbilityBuilder, Ability} from '@casl/ability'
-import {loadCurrentUser} from 'lib/users'
 import {intersection, isString} from 'lodash'
+import {loadCurrentViewerRoles} from '../lib/viewer'
 
 type Actions = 'manage' | 'upload'
 type Subjects = 'Video' | 'all'
+export type Roles = 'admin' | 'editor' | 'publisher'
```
Comment (joelhooks) [inline src/server/ability.ts:7]: this gives type checking which is nice, only drawback was having to cast in the tests

File: src/server/ability.ts [hunk 3]
```diff
@@ -1,42 +1,42 @@
...
-  const user = await loadCurrentUser(token)
-  return defineAbilityFor(user)
+  const viewerRoles = await loadCurrentViewerRoles(token)
+  return defineAbilityFor(viewerRoles)
 }
 
-function defineAbilityFor(user: any) {
-  // AbilityBuilder https://casl.js.org/v5/en/guide/define-rules#ability-builder-class
+/**
+ * @see {@link https://casl.js.org/v5/en/guide/define-rules#ability-builder-class|AbilityBuilder}
```
Comment (joelhooks) [inline src/server/ability.ts:16]: jsdoc is pretty sweet

you can actually replace TS with it which is interesting

File: src/server/ability.ts [hunk 4]
```diff
@@ -1,42 +1,42 @@
...
+export function includesRoles(
+  roles: Roles[] = [],
+  rolesToCheck: Roles | Roles[] = [],
+) {
   if (isString(rolesToCheck)) {
-    rolesToCheckArray = [rolesToCheck]
-  } else {
-    rolesToCheckArray = rolesToCheck
+    rolesToCheck = [rolesToCheck]
   }
```
Comment (joelhooks) [inline src/server/ability.ts:39]: there are some various arguments (ha) for not overwriting like this, but in this case the extremely marginal risk seems fine

File: src/server/ability.ts [hunk 5]
```diff
@@ -1,42 +1,42 @@
...
     can('upload', 'Video')
   }
 
   return build()
 }
 
-export function hasRoles(user: any, rolesToCheck: string | string[]) {
-  // ensure rolesToCheck is an array
-  let rolesToCheckArray: string[]
+export function includesRoles(
```
Comment (joelhooks) [inline src/server/ability.ts:33]: I went with includes because it's similar to includes and since it's typed now it's pretty true

Comment (jbranchaud) [reply src/server/ability.ts:33]: I like this name better 💯

**Review by jbranchaud — APPROVED** [review]

**Pull request merged**